src/modules/bulletproofs/inner_product_impl.h

```diff
@@ -0,0 +1,848 @@
+/**********************************************************************
+ * Copyright (c) 2018 Andrew Poelstra                                 *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef SECP256K1_MODULE_BULLETPROOF_INNER_PRODUCT_IMPL
+#define SECP256K1_MODULE_BULLETPROOF_INNER_PRODUCT_IMPL
+
+#include "group.h"
+#include "scalar.h"
+
+#include "modules/bulletproofs/main_impl.h"
+#include "modules/bulletproofs/util.h"
+
+#define POPCOUNT(x)	(__builtin_popcountl((unsigned long)(x)))  /* TODO make these portable */
+#define CTZ(x)	(__builtin_ctzl((unsigned long)(x)))
+
+/* Number of scalars that should remain at the end of a recursive proof. The paper
+ * uses 2, by reducing the scalars as far as possible. We stop one recursive step
+ * early, trading two points (L, R) for two scalars, which reduces verification
+ * and prover cost.
+ *
+ * For the most part, all comments assume this value is at 4.
+ */
+#define IP_AB_SCALARS	4
+
+/* Bulletproof inner products consist of the four scalars and `2[log2(n) - 1]` points
+ * `a_1`, `a_2`, `b_1`, `b_2`, `L_i` and `R_i`, where `i` ranges from 0 to `log2(n)-1`.
+ *
+ * The prover takes as input a point `P` and scalar `c`. It proves that there exist
+ * scalars `a_i`, `b_i` for `i` ranging from 0 to `n-1`, such that
```

Comment:
This is still somewhat confusing with the index 0 of `a` and `b`.

Reply:
Agreed, at this point in the comment I should 1-index.